File: java/src/main/java/core/PhoneBook.java
package core;

import java.util.List;
import java.util.stream.Collectors;

public class PhoneBook {

    public List<Contact> getContacts() {
        return contacts;
    }

    private List<Contact> contacts;

    public PhoneBook(List<Contact> contacts) {
        this.contacts = contacts;
    }

    public String reverseLookup(String phoneNumber) {
        return contacts.stream()
                .filter(contact -> phoneNumber.equals(contact.getPhoneNumber()))
                .findFirst()
                .map(Contact::getName)
                .orElse(null);
    }

    public String lookup(String name) {
        return contacts.stream()
                .filter(contact -> name.equalsIgnoreCase(contact.getName()))
                .findFirst()
                .map(Contact::getPhoneNumber)
                .orElse(null);
    }

    public List<Contact> findAllContact(String searchParameter) {
        if (searchParameter.isBlank()) {
            return contacts;
        }
        return contacts.stream()
                .filter(contact -> contact.getName().contains(searchParameter) ||
                        contact.getPhoneNumber().contains(searchParameter))
                .collect(Collectors.toList());
    }

    public boolean addContact(Contact newContact) {
        if (contacts.contains(newContact)) {
            return false;
        }
        contacts.add(newContact);
        return true;
    }

    public boolean removeContact(Contact contactToRemove) {
        if (contacts.contains(contactToRemove)) {
            return false;
        }
        contacts.remove(contactToRemove);
        return true;
    }

}

File: js/prod/PhoneBook.ts
import Contact from "./Contact";

export default class PhoneBook {

    public get getContacts() :Array<Contact> {
        return this.contacts;
    }

    public constructor(private contacts: Array<Contact>) {
    }

    public reverseLookup( phoneNumber:string):string {
        return this.contacts
                .filter(contact => phoneNumber == contact.getPhoneNumber)
                .map(contact => contact.getName)
                .shift();
    }

    public lookup(name : string) :string {
        return this.contacts.filter(contact => name == contact.getName)
                .map(contact => contact.getPhoneNumber)
                .shift();
    }

    public findAllContact (searchParameter: string): Array<Contact> {
        if (searchParameter == "") {
            return this.contacts;
        }
        return this.contacts
                .filter(contact => contact.getName.includes(searchParameter) ||
                        contact.getPhoneNumber.includes(searchParameter));
    }

    public addContact( newContact:Contact) : boolean{
        if (this.contacts.includes(newContact)) {
            return false;
        }
        this.contacts.push(newContact);
        return true;
    }

    public removeContact( contactToRemove: Contact) :boolean {
        if (this.contacts.includes(contactToRemove)) {
            return false;
        }
        this.contacts.filter(x => x != contactToRemove);
        return true;
    }

}